src/Util/Visitor/VariableCollectionVisitor.php

```diff
@@ -0,0 +1,287 @@
+        //       actually have a return type??
+        if ($expr instanceof Expr\FuncCall) {
+            $func = (string) $expr->name;
+            $reflection = new \ReflectionFunction($func);
+
+            // do not try and find out return type for non-internal functions
+            if (false === $reflection->isInternal()) {
+                return;
+            }
+
+            // in the case that no return type was provided, just return
+            // "mixed".
+            return $reflection->getReturnType() ?: $this->createUnknownReflectionType();
+        }
+
+        if ($expr instanceof Expr\ArrayDimFetch) {
+            return $this->createUnknownReflectionType();
+        }
+
+        if ($expr instanceof Expr\Array_) {
+            return $this->createReflectionTypeFromString('array');
+        }
+
+        if ($expr instanceof Node\Scalar) {
+            switch (get_class($expr)) {
+                case Node\Scalar\DNumber::class:
+                    return $this->createReflectionTypeFromString('float');
+
+                case Node\Scalar\LNumber::class:
+                    return $this->createReflectionTypeFromString('integer');
+
+                case Node\Scalar\String_::class:
+                    return $this->createReflectionTypeFromString('string');
+
+                case Node\Scalar\Encapsed::class:
+                case Node\Scalar\MagicConst::class:
+                case Node\Scalar\EncapsedStringPart::class:
+                    // TODO: ???
+                    return $this->createUnknownReflectionType();
+                default:
+                    throw new \RuntimeException(sprintf(
+                        'Do not know scalar type "%s"', get_class($expr)
+                    ));
+            }
+        }
+
+        throw new \RuntimeException(sprintf(
+            'Could not determine type from expression for node of type "%s"',
+            get_class($expr)
+        ));
+    }
+
+    /**
+     * Create the a type relative to the current reflection class from
+     * a php-parser name-node.
+     */
+    private function createTypeFromNameNode(Name $name)
+    {
+        $docType = (new FindTypeFromAst())->__invoke(
+            $name,
+            $this->reflection->getLocatedSource(),
+            $this->reflection->getNamespaceName()
+        );
+
+        return $this->createReflectionTypeFromDocType($docType);
+    }
+
+    private function createReflectionTypeFromString(string $type = null)
+    {
+        $type = $type ?: 'mixed';
+
+        return $this->createReflectionTypeFromDocType(
+            $this->typeResolver->resolve($type)
+        );
+    }
+
+    private function createReflectionTypeFromDocType(Type $type): ReflectionType
+    {
+        return ReflectionType::createFromType($type, false);
+    }
+
+    private function createUnknownReflectionType()
+    {
+        return $this->createReflectionTypeFromString('mixed', false);
```

Comment:
Should be `null`, IMO. Adding `mixed` is a very specific thing, plus nothing prevents a `class mixed {}` to exist.